File: 3583-count-special-triplets.go
//go:build ignore

package main

import "fmt"

func specialTriplets(nums []int) int {
	mod := 1_000_000_007
	n := len(nums)
	prefix := make([]int, n)
	count := make(map[int]int)
	for j, num := range nums {
		prefix[j] = count[num*2]
		count[num]++
	}
	res := 0
	count = make(map[int]int)
	for j := n - 1; j >= 0; j-- {
		num := nums[j]
		suffix := count[num*2]
		res = (res + (prefix[j] * suffix)) % mod
		count[num]++
	}

	return res
}

func main() {
	fmt.Printf("Output: %d, expected: 1\n", specialTriplets([]int{6, 3, 6}))
	fmt.Printf("Output: %d, expected: 1\n", specialTriplets([]int{0, 1, 0, 0}))
	fmt.Printf("Output: %d, expected: 2\n", specialTriplets([]int{8, 4, 2, 8, 4}))
}

File: 3583-count-special-triplets.py
from collections import defaultdict


class Solution:
    def specialTriplets(self, nums: list[int]) -> int:
        '''
        We can use a modified of prefix sum and suffix sum whereby we calculate for each number at j,
        how many occurrences of number * 2 are there before it (prefix sum)

        Next, we loop through again but from the back and do the same check and multiply the 2 values at each index and sum them up
        '''
        MOD = 10 ** 9 + 7
        n = len(nums)
        prefix = [0] * n
        count = defaultdict(int)
        for j, num in enumerate(nums):
            prefix[j] = count[num * 2]
            count[num] += 1

        res = 0
        count = defaultdict(int)
        for j in range(n - 1, -1, -1):
            num = nums[j]
            suffix = count[num * 2]
            res = (res + (prefix[j] * suffix)) % MOD
            count[num] += 1

        return res


sol = Solution()
print(f'output: {sol.specialTriplets([6, 3, 6])}, expected: 1')
print(f'output: {sol.specialTriplets([0, 1, 0, 0])}, expected: 1')
print(f'output: {sol.specialTriplets([8, 4, 2, 8, 4])}, expected: 2')
print(f'output: {sol.specialTriplets([8, 4, 8, 4, 8])}, expected: 4')
